xcm/src/v0/multi_location.rs

```diff
@@ -44,77 +44,424 @@ use parity_scale_codec::{self, Decode, Encode};
-	/// A relative path comprising 8 junctions.
-	X8(Junction, Junction, Junction, Junction, Junction, Junction, Junction, Junction),
+#[derive(Clone, Eq, PartialEq, Ord, PartialOrd, Debug)]
+pub struct MultiLocation {
+	parents: u8,
+	interior: Junctions,
+}
+
+impl Default for MultiLocation {
+	fn default() -> Self {
+		Self::empty()
+	}
+}
+
+impl Encode for MultiLocation {
+	fn encode_to<T: Output + ?Sized>(&self, dest: &mut T) {
+		// 255 items max, but we cannot return an error, so we truncate silently.
+		let parents = self
+			.parents
+			.min(MAX_MULTILOCATION_LENGTH.saturating_sub(self.interior.len()) as u8);
+		let count = parents.saturating_add(self.interior.len() as u8);
+		dest.push_byte(count);
+		for _ in 0..parents {
+			dest.push_byte(0);
+		}
+		for j in &self.interior {
+			j.encode_to(dest)
+		}
+	}
+}
+
+impl Decode for MultiLocation {
+	fn decode<I: Input>(input: &mut I) -> Result<Self, parity_scale_codec::Error> {
+		let count = input.read_byte()?;
+		let mut parents = 0u8;
+		let mut interior = Junctions::Null;
+		for i in 0..count {
+			let b = input.read_byte()?;
+			if b == 0 {
+				if i > parents {
+					// this would imply that there has been at least one non-parent junction, which
+					// further implies that this is non-canonical. we bail.
+					return Err(parity_scale_codec::Error::from("Parent after non-Parent forbidden"))
+				}
+				parents = parents.saturating_add(1);
+			} else {
+				let mut bytes = [b].to_vec();
+				loop {
+					if let Ok(junction) = Junction::decode(&mut &bytes[..]) {
+						interior = interior
+							.pushed_with(junction)
+							.map_err(|_| parity_scale_codec::Error::from("Interior too long"))?;
+						break
+					} else {
+						// a single byte may not be enough to decode a `Junction`, so keep reading
+						// the next byte until a `Junction` can successfully be decoded
+						bytes.push(input.read_byte()?);
+					}
+				}
+			}
+		}
+		Ok(MultiLocation { parents, interior })
+	}
 }
 
 /// Maximum number of junctions a `MultiLocation` can contain.
-pub const MAX_MULTILOCATION_LENGTH: usize = 8;
+pub const MAX_MULTILOCATION_LENGTH: usize = 255;
+
+impl MultiLocation {
+	/// Creates a new `MultiLocation`, ensuring that the length of it does not exceed the maximum,
+	/// otherwise returns `Err`.
+	pub fn new(parents: u8, junctions: Junctions) -> result::Result<MultiLocation, ()> {
+		if parents as usize + junctions.len() > MAX_MULTILOCATION_LENGTH {
+			return Err(())
+		}
+		Ok(MultiLocation { parents, interior: junctions })
+	}
+
+	/// Creates a new `MultiLocation` with 0 parents and a `Null` interior.
+	pub const fn empty() -> MultiLocation {
+		MultiLocation { parents: 0, interior: Junctions::Null }
+	}
+
+	/// Creates a new `MultiLocation` with the specified number of parents in the `P` const generic
+	/// parameter and a `Null` interior.
+	pub const fn with_parents_const<const P: u8>() -> MultiLocation {
```

Comment:
It is absolutely nuts that this function actually works. I am wondering whether I should call this `with_parents` and remove the existing implementation of that method altogether...

Reply:
will rustc monomorphizes const generics as well? in that case for each use we get a copy of this function lying around, potentially bloating the wasm code. 

And why does this need to be const in the first place?

Reply:
We can't accept a `Junctions` as a parameter in a const eval context, which is why you'll see no const constructors here that do. The reasoning is because `Junctions` contains a variant with a `Vec<u8>`, which is a type that will run a destructor when it goes out of scope, and running destructors are prohibited in a const eval context.

Not too sure about monomorphization, but I can find that one out.

Reply:
Looks like it does monomorphize const generics: https://github.com/rust-lang/rfcs/blob/master/text/2000-const-generics.md

Reply:
In practical terms however, I would expect people to only call this with numbers ranging from 1 to 4 to denote their distance from the relay chain. I would not expect to see them iterating from 0 to 255 just to test all possible `u8` digits to fit into this function.

Reply:
> In practical terms however, I would expect people to only call this with numbers ranging from 1 to 4 to denote their distance from the relay chain. I would not expect to see them iterating from 0 to 255 just to test all possible `u8` digits to fit into this function.

This sounds reasonable, but I didn't receive my answer in the first place that when do we need `with_parents_const` itself?

Reply:
I think you can remove this function overall https://github.com/paritytech/polkadot/pull/3404#discussion_r684411842